// ng2-filter-pipe/__ivy_ngcc__/dist/src/ng2-filter.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Created by vadimdez on 28/11/2016.
 */
var core_1 = require("@angular/core");
var ng2_filter_pipe_1 = require("./ng2-filter.pipe");
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./ng2-filter.pipe');
var Ng2FilterPipeModule = (function () {
    function Ng2FilterPipeModule() {
    }
Ng2FilterPipeModule.ɵfac = function Ng2FilterPipeModule_Factory(t) { return new (t || Ng2FilterPipeModule)(); };
Ng2FilterPipeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng2FilterPipeModule });
Ng2FilterPipeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2FilterPipeModule, { declarations: [ɵngcc1.Ng2FilterPipe], exports: [ɵngcc1.Ng2FilterPipe] }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng2FilterPipeModule, [{
        type: core_1.NgModule,
        args: [{
                declarations: [ng2_filter_pipe_1.Ng2FilterPipe],
                exports: [ng2_filter_pipe_1.Ng2FilterPipe]
            }]
    }], function () { return []; }, null); })();
    return Ng2FilterPipeModule;
}());
/** @nocollapse */
Ng2FilterPipeModule.ctorParameters = function () { return []; };
exports.Ng2FilterPipeModule = Ng2FilterPipeModule;
